source/credentials_provider_cognito.c

```diff
@@ -0,0 +1,925 @@
+        aws_parse_credentials_from_aws_json_object(user_data->allocator, credentials_entry, &credentials_parse_options);
+    if (user_data->credentials == NULL) {
+        goto done;
+    }
+
+    result = AWS_OP_SUCCESS;
+
+done:
+
+    aws_json_value_destroy(response_document);
+
+    if (result != AWS_OP_SUCCESS) {
+        aws_raise_error(AWS_AUTH_PROVIDER_PARSER_UNEXPECTED_RESPONSE);
+    }
+
+    return result;
+}
+
+static void s_on_retry_ready(struct aws_retry_token *token, int error_code, void *user_data) {
+    (void)token;
+    struct cognito_user_data *provider_user_data = user_data;
+
+    if (error_code != AWS_ERROR_SUCCESS) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p): Cognito credentials provider retry task failed: %s",
+            (void *)provider_user_data->provider,
+            aws_error_str(error_code));
+        s_finalize_credentials_query(user_data, error_code);
+        return;
+    }
+
+    s_user_data_reset(provider_user_data);
+
+    struct aws_credentials_provider_cognito_impl *impl = provider_user_data->provider->impl;
+
+    impl->function_table->aws_http_connection_manager_acquire_connection(
+        impl->connection_manager, s_on_connection_setup_fn, provider_user_data);
+}
+
+static void s_on_stream_complete_fn(struct aws_http_stream *stream, int error_code, void *user_data) {
+
+    struct cognito_user_data *provider_user_data = user_data;
+    struct aws_credentials_provider_cognito_impl *impl = provider_user_data->provider->impl;
+
+    int http_response_code = 0;
+    impl->function_table->aws_http_stream_get_incoming_response_status(stream, &http_response_code);
+
+    if (http_response_code != 200) {
+        error_code = AWS_AUTH_CREDENTIALS_PROVIDER_HTTP_STATUS_FAILURE;
+    }
+
+    impl->function_table->aws_http_stream_release(stream);
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+        "(id=%p): GetCredentialsForIdentity call completed with http status %d",
+        (void *)provider_user_data->provider,
+        http_response_code);
+
+    if (http_response_code == AWS_HTTP_STATUS_CODE_200_OK) {
+        aws_retry_token_record_success(provider_user_data->retry_token);
+
+        if (s_parse_credentials_from_response(provider_user_data) == AWS_OP_SUCCESS) {
+            s_finalize_credentials_query(user_data, AWS_ERROR_SUCCESS);
+            return;
+        }
+
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p): Cognito credentials provider failed to parse GetCredentialsForIdentity response",
+            (void *)provider_user_data->provider);
+
+        error_code = AWS_AUTH_PROVIDER_PARSER_UNEXPECTED_RESPONSE;
+    }
+
+    /* Success path is done, error-only from here on out */
+
+    /* Unsure if this should be unconditional or a function of status code. STS does this unconditionally. */
+    impl->function_table->aws_http_connection_close(provider_user_data->connection);
```

Comment:
is aws_http_connection_manager_release_connection() still supposed to be called in this case?

Reply:
Should be called in s_user_data_reset